### DISCUSSION TITLE
Steps towards version 1.0

### DISCUSSION THREAD

**Discussion by jsmariegaard** [DHI/modelskill#185] [Ideas]
We are currently re-factoring this library making it ready for version 1.0. 

## Goals

- [x] Ready to become the validation engine in MOOD (circular stats, dual rose plot)
- [x] Support of more model types (e.g. also machine learning models, WWTP) 
- [x] Comparer should have a flexible index - e.g. allow for additional axis 
    - Prepare for new geometry types (e.g. vertical profile data or raster data) 
    - Prepare for (or support) forecast skill assessment (forecast data would require lead_hour axis as well as time)
- [x] Simplified and consistent api and workflow (e.g. remove Connector, no filtering on plotting functions)
- [x] Simple to make Comparer from already matched data 
- [x] Simplified internal structure for easier extension and increased maintainability
- [x] Load/save comparisons to file 
- [x] Simpler to filter/aggregate using auxiliary data (co-variates) and custom meta data (attrs)
- [x] Better quantity/unit/iteminfo handling (don't force users to use EUM if not dfs file)
- [x] Should be easy to change names and basic properties after comparison (e.g. if name doesn't look good on plots, it should be possible to change without re-doing comparison from start)

## New API

If the modelresult and observation geometry types can easily be inferred (e.g. dfsu or point data): 

     import modelskill as ms
     mr1 = ms.model_result(...)
     mr2 = ms.model_result(...)
     o1 = ms.observation(...)
     o2 = ms.observation(...)
     cc = ms.compare([o1, o2], [mr1, mr2])

If the user wants to be more explicit (or geometry types cannot easily be inferred): 

     import modelskill as ms
     mr1 = ms.GridModelResult(...)
     mr2 = ms.DfsuModelResult(...)
     o1 = ms.PointObservation(...)
     o2 = ms.TrackObservation(...)
     cc = ms.compare([o1, o2], [mr1, mr2])

If data has already been matched (e.g. resides in a single dfs0/dataframe): 

    import modelskill as ms
    cmp = ms.from_matched(df, obs_item, mod_item, ...)

If the modelresults and observations are specified in a configuration file: 

    import modelskill as ms
    cc = ms.from_config("conf.xlxs")

### Week 3

-  See documentation section below (prioritize docstrings)
- repr
    - [ ] out.append(f" Model: {model}, rmse={self.score()[model]:.3f}") - remove score?
    - [ ] Improved repr strings, #378 
    - [ ] SkillGrid repr makes user think it is a xarray! (confusing)
- remove "return getattr" from `__getattr__` in both SkillTable and SkillGrid? #391 
- ask for testing (DACA) 

## Documentation

- [ ] Document recently added features
- [x] DummyModelResult page
- [x] rename API to API Reference (like pandas and xarray), #363
- [x] Make section: "User Guide" (like pandas and xarray), #363
     - Selecting/filtering data
- [ ] Make section "Examples", #341
- [ ] Decide on **tagline** and use it consistently!
- [ ] Update readme (remove altimetry, add nicer plots) - links to notebooks are no longer valid (after renaming notebooks)
- [ ] Remove Taylor Diagram from most of the notebooks (too difficult to understand)
- [ ] docstrings on attributes, like `data` and `plot` (make them appear in the documentation)
- [ ] docstring missing: Comparer.remove_bias
- [ ] docstring examples in ...
    - [ ] match()
- [ ] docstrings: consistent way of showing multiple arg types. Now: [str, int], (str, int) and "str or int" - should be the latter
- [ ] add "Trajectory" to description of Track
- [ ] Hide subsections like "See Also" in right pane
- [ ] Add to issues: Bug: It is not possible to set user-defined default metrics for directional data (will be overridden); it can still be set manually in skill() 
- [ ] (how to document api of timeseries.plot (it does not appear anywhere in the docs)!, (partly handled by #363 ))
- [ ] (Group notebooks so everything with altimetry is in a subfolder (not to scare non-marine users away))

## Bugs/minor

- [ ] Prematched_with_auxiliary.ipynb fails! (scatter plot on cmp3) (still fails 2024-1-4) -> now added as issue

## Testing

- [ ] multi-variable SkillTable with wind speed and wind direction (different metrics)
- [x] Protected names: "Observation" is probably not allowed - test (Result: "Observation" is a valid obs name - it is used in several tests)
- [x] Verify that Comparer/ComparerCollection: **rename** method allows the user to rename any column (model, obs, aux)
- [ ] Verify that #220 has been solved
- [ ] check problem with combining comparers in WR case notebook

## January, 2024, post-release, but soon

- [ ] Handling of spatial interpolation (new spatial_interp_method arg?)
- [ ] Timeseries plot time axis appearance not nice if non-equidistant calendar, #347 
- [ ] Color handling: each model should have a unique color which should be assigned when creating comparer. (re-surface color property?)
- [ ] attrs on model could be really useful for calibration (especially if calibrating several params over many runs); allow user to provide attrs on MR and filter/groupby on model attrs
- [ ] DfsuModelResult and GridModelResult misses properties (names, quantities, time)
- [ ] Check how many changes are needed to handle 2d or 3d index (e.g. time,z or time,lead_hour or time,x,y) (test by xr.expand_dims)
- [ ] allow from_matched to take xr.Dataset as input
- [ ] cc.plot.temporal_coverage() 
- [ ] How do we make it simple to change quantity (e.g. unit) (Make quantity mutable to it is easy to change long_name etc; postpone)
- [ ] **SkillArray.plot should have a plot.map and/or plot.explore function**
- [ ] SkillGrid could have `quantity` property for consistency
- [ ] sel() method should use part of e.g. model name as long as it is unique
- [ ] sel() method should allow use of "*" (like in mikeio)
- [ ] drop() method to drop obs/mod/variable (reverse sel)
- [ ] (API: should comparers have both query() and where()? )
- [ ] Scatter ax argument in case of multiple models (=multiple plots)
- [ ] cc.plot: box plot is missing! (currently only Comparer supports box plot)
- [ ] cc.plot: qq and residual_hist missing! (currently only in Comparer - hence inconsistent)
- [ ] (Add name argument in from_matched() and/or match(), already so???)
- [ ] Methods in SkillTable could be moved to SkillMixIn (e.g. sort_values) and then use in both SkillTable and SkillArray
- [ ] point/grid should take .nc files similar to GridModelResult
- [ ] Time interpolation of directional data - avoid interpolating over 360-0.
- [ ] Check spatial match when comparing point to point (new spatial_tolerance arg in match?)
- [ ] add z to _to_long_dataframe() if not None or np.nan

## Refactoring

* ItemSelection when reading files: create _utils module in _timeseries: Refactor the way we select/validate items. It is now quite similar in Observation and ModelResults - track is a bit different (also has x and y item), but the rest is the same. This should be refactored out. Postponed from #304 
- Comparer/Timeseries: A way to bypass dataset validation if a result of filtering
- groupby code in skill methods

## Considerations/thoughts

* How to handle weighted metrics (sample_weights)? (use "weights" in inspect.signature(metric).parameters to find out if sample_weights should be applied in a specific)
* Should we allow np.nan in Comparer (maybe only drop if necessary)
* How should empty comparers be handled (e.g. upon filtering)? 
* What to do with empty comparers? Allowed? Not-allowed? Throw error? (e.g. if no overlapping data in matching)
* Create example file structure (how we suggest you organize your files)

### Naming

* Should we rename swap_level in skill table? 
* Should we use name "Dfsu", "FlexibleMesh" or "Unstructured"?
* Consistent naming: mod_ or model_ or mr or modelresults
* Should we rename items to avoid spaces, commas etc - makes it difficult to use query() etc
* Use variable or quantity (or both)

## Long-list, postponed, 2024

* isel() method? #374 
* SkillGrid could have to_geodataframe() like SkillTable
* Comparer/ComparerCollection plotting with multiple models
	- cmp.plot.scatter() behaviour - different markers, BW vs colorscale, multiple axes?
* Forecast skill: Interpolation only in forecast direction, 2024
* Idea: Polygon Skill - skill aggregated over a basin (assign observations to group ...)
* How to include non-metric analytics (max, std etc) in skill table? (could add as an example, if SkillTable is added to root namespace)
* (AggregatedSkill: Change data container to xarray dataset, postponed)
* (Remove Connector when config, plots etc have been moved. Update tests and examples accordingly. ) wait until version 1.1
* (Should the skill() method return a SkillArray instead of a SkillTable if metrics is a str?, postponed/dropped)
* re-establish mean_skill_points() (currently disabled as no longer correct)
* Should we allow users to merge two SkillTable objects? (e.g. if one contains weighted and another not)
* sk.style of new/own metrics

## Done

- [x] Check for use of "id" where "idx" is really meant
- [x] Make a new notebook with WWTP example, #336 
- [x] API split into many smaller files, grouped
- [x] Perf: use categorical vars instead of strings to identify obs/model in long dataframe view in ComparerCollection. -> faster groupby, #351
- [x] Introduce new spatial_interp argument? 
- [x] Skill groupby: sort=False, #376 
- [x] Bug: Skill groupby - set track observation x, y to NaN, #376
- [x] Re-introduce cmp.to_dataframe() method after renaming existing internal func (long format), #376 
- [x] Release final beta, Monday, January 8
- [x] API: Make SkillTable accessible from root name space and init should take both df and xr.ds, #372
- [x] API: Remove more properties: #373 
     - Collection has a few properties that are undocumented and maybe should be removed or made private? name, unit_text, n_comparers, n_observations (maybe keep for consistency)? 
     - geometry property (shapely) on obs objects? (or rename to to_shapely() )
- [x] **API: plotting.scatter take a skill_df argument which is inconsistent with cc.plot.scatter**
    - [x] scatter plot should NOT take skill_df but a dict (see e.g. format_skill_df) #369 
- [x] API: Scatter plot argument list are slightly different in Comparer vs ComparerCollection! Difference: "norm" and "ax". #370
- [x] ms.from_config still returns connector! Should return ComparisonCollection now!, #366 
- [x] **Bug: rename Observation name does not work in comparer!**, #368
- [x] **Protected names: "time", "x", "y", "z", "Observation" can not be used for naming observations and modelresults; add check (also to rename method)**, #368
- [x] API: rename _variable_ to _quantity_ (e.g. in sel() method), #365 
- [x] Bug: fix random colors in observations (decide on fixed color for observation, unless defined by user), #367
- [x] skill() groupby attrs, #351
- [x] API: Make Comparer/Collection to_dataframe method internal: _to_long_dataframe(), #351
- [x] **Bug: multi-model scatter plots show same skill table!!** 
- [x] API: remove more properties from classes, e.g. mod, #354 
- [x] skill() groupby Month-in-Year etc, #352
- [x] Should we remove "weight" as an attribute on observation/comparer (only used in a single method: mean_skill() )... ? #342
- [x] API: Re-establish weighted observations? (currently broken also color) or should it just be on mean skill - see consideration below, #342
- [x] Bug: box plot shows x,y, z as if they were variables 
- [x] Any risks/problems with "observation" which is both the name of a root-namespace function and a module!?
- [x] API: consistency: end (comparer) or end_time (timeseries): TODO: remove except on cc (rename to end_time), #353
- [x] API: consistency: field_names (SkillGrid) or metrics (SkillTable)
- [x] Rename SpatialSkill to GriddedSkill or SkillGrid, #315 
- [x] Allow compare to return Comparer if observation is str (more consistent with from_matched)
- [x] Make sure all tests and examples etc use "import modelskill as ms"
- [x] Remove all usage of Connector
- [x] Make a new notebook with urban drainage example, #274 
- [x] Make a new notebook with water resource example (river network), #249 
- [x] AggregatedSkill: split in a dataset and a dataarray part, so that you do not need to specify "metric" in the plot functions, #309 
- [x] ms.observation() function similar to ms.model_result(), #328 
- [x] CHGA: score() function should return a dictionary if multiple models (not clear enough which score belongs to which model) - already so for collection, #330
- [x] Comparer.plot: multiple models should lead to multiple plots (for now - later we can add options), #337 
- [x] hist plot still has `model` argument -> return multiple plots instead, #337 
- [x] mean_skill_points() no longer correct! -> out-comment for now
- [x] SkillGrid sel() method to avoid "model" argument in plot() method, #340 
- [x] remove_bias() should return a new Comparer object instead of modifying self
- [x] Circular statistics (e.g. wave direction), #284 
- [x] ItemInfo to long_name and unit (i.e. if input is dfs file make sensible conversion to long_name + unit AND change plots/tables to generate labels based on long_name and unit instead of itemInfo) #191 
- [x] GridSkill: split in a dataset and a dataarray part, so that you do not need to specify "metric" in the plot functions 
- [x] Clean-up use of types, protocols and base classes - should be consistent across the library and easy to understand and work with in the future, @ecomodeller  
- [x] Clean-up: e.g. rename all data fields to .data. 
- [x] Remove multi-item modelresults. 
- [x] Change class structure of ModelResults such that they Resemble Observations: PointModelResult, TrackModelResult etc. 
- [x] Create new internal data structure .data for SingleObsComparer and other Comparers based on xarray. 
- [x] Implement sel() method. 
- [x] Change internal data structure of ModelResults and Observations to use xarray dataset (make it simpler to carry over meta data to Comparer)
- [x] Allow both ModelResults and Observations to carry aux data (aux_item argument)
     - Example: upon skill assessment of wave period (Tp) I want to make it conditional on wave height (Hm0). The ModelResult has that information and modelskill could extract Hm0 from the dfsu file together with Tp. Or maybe the measurement file has that information and we would like to get it from there. (The user would perhaps like to assign a custom name to the aux variable e.g. "obs_Hm0"), #304 
- [x] Allow custom metadata from Observations (and maybe ModelResult) in an attrs container to be carried forward to the Comparer. 
     - Example: in data assimilation cases, I would like to be able to distinguish between "DA" and "validation" stations. In order cases, it could be different "tanks" in a WWTP or offshore/nearshore stations, #291 
generate labels based on long_name and unit instead of itemInfo) #191 
- [x] Rich compare() method (removing the need for the connector) #180 
- [x] Move from_config() out of Connector and into new module configuration.py or configurator.py (how to handle to_config? (method in root name space and/or method on ComparerCollection?)) 
- [x] Move plot_temporal_coverage() and plot_observation_positions() to root name space #187
- [x] Deprecate Connector by issuing deprecation warnings. 
- [x] **Load/save** functionality in Comparer and ComparerCollection #178 
- [x] New function ms.from_matched(data, obs_item, mod_item) #192 
- [x] New function validate_comparer_dataset (to check the consistency of the xarray dataset that constitutes Comparer)
- [x] Rename package to modelskill
- [x] Update all examples, notebooks and tests to use new api (import modelskill as ms, ...)
- [x] Use new logo #184 
- [x] Make pre-release of modelskill to pypi to reserve the name 
- [x] Check how many changes are needed to handle data without time axis? #293 
- [x] Handling of circular variables (also plots etc - grid at 0, 90, 180, 270 and 360)
- [x] Dual-rose plot for comparison of two circular data sets, @daniel-caichac-DHI 
- [x] Loosen requirement in Comparer that dataset must have a DateTimeIndex, #293

**Comment by Hendrik1987** [comment]
@jsmariegaard : This seems very exciting!
Are you planning on some pre-release of 1.0? Would be keen to try it out on a project.
Cheers!

**Comment by jsmariegaard** [comment]
@Hendrik1987, the first alpha is already on PyPI: https://pypi.org/project/modelskill/ but this is mostly to reserve the name. Proper pre-releases will be made soon after the summer in August/September.
